src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,533 @@
+
+                /* Fetch memlock from /proc/self/fdinfo using map fd */
+                r = get_memlock_by_fd(fd, &memlock);
+                if (r < 0)
+                        continue;
+
+                r = json_build(&v, JSON_BUILD_OBJECT(
+                                        JSON_BUILD_PAIR_STRING("name", map_info.name),
+                                        JSON_BUILD_PAIR_UNSIGNED("type", map_info.type),
+                                        JSON_BUILD_PAIR_UNSIGNED("memlock", memlock)));
+                if (r < 0) {
+                        log_error_errno(r, "Failed to build map JSON object, ignoring: %m");
+                        continue;
+                }
+
+                r = json_variant_append_array(&map_list, v);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to append to json variant array for map, ignoring: %m");
+                        continue;
+                }
+        }
+
+        *ret = TAKE_PTR(map_list);
+
+        return 0;
+}
+
+
+static int fetch_bpf_programs(JsonVariant **ret) {
+        _cleanup_hashmap_free_free_ Hashmap *map_name_by_id = NULL;
+        uint32_t id = 0;
+        JsonVariant *prog_list = NULL;
+        int r;
+
+        /* Iterate through all attached BPF progs*/
+        id = 0;
+        while (sym_bpf_prog_get_next_id(id, &id) >= 0) {
+                struct bpf_prog_info prog_info = {};
+                uint32_t prog_info_len = sizeof(prog_info);
+                uint32_t nr_map_ids = 0;
+                uint64_t memlock = 0;
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char *prog_name = NULL, *prog_tag = NULL;
+                _cleanup_free_ uint32_t *prog_map_ids = NULL;
+                _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+                _cleanup_strv_free_ char **prog_map_names = NULL;
+
+                /* Grab the FDs using their IDs*/
+                fd = sym_bpf_prog_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error_errno(errno, "Can't get fd for prog id %u, ignoring: %m", id);
+                        continue;
+                }
+
+                /* Fetch bpf_prog_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                if (r < 0) {
+                        log_error_errno(r, "Can't get bpf prog info for fd %i, ignoring: %m", fd);
+                        continue;
+                }
+
+                /* If there are maps associated with this prog, fetch bpf_prog_info again
+                 * to get a list of map ids */
+                nr_map_ids = prog_info.nr_map_ids;
+                if (nr_map_ids > 0) {
+                        prog_map_ids = new(uint32_t, nr_map_ids);
+                        if (!prog_map_ids)
+                                return log_oom();
+
+                        zero(prog_info);
+                        prog_info = (struct bpf_prog_info) {
```

Comment:
It doesn't make sense to zero the struct, if it's about to be immediately overridden anyway.